src/dataflow/source/kinesis.rs

```diff
@@ -46,11 +64,98 @@ where
+        };
+        let shard_iterator_output = block_on(get_shard_iterator(&client, &shard, stream_name));
+        let mut shard_iterator = shard_iterator_output.shard_iterator.unwrap();
+
+        move |cap, output| {
+            // If reading from Kinesis takes more than 10 milliseconds,
+            // pause execution and reactivate later.
+            let timer = std::time::Instant::now();
+
+            let mut get_records_response = get_records_wrapper(&client, &shard_iterator);
+            // When the next_shard_iterator is null, the shard has been closed and the
+            // requested iterator does not return any more data.
+            while let Some(iterator) = get_records_response.next_shard_iterator {
+                if let Some(millis) = get_records_response.millis_behind_latest {
+                    if millis == 0 {
+                        // This activation does the following:
+                        //      1. Ensures we poll Kinesis more often than the eviction timeout (5 minutes)
+                        //      2. Proactively and frequently reactivates this source, since we don't have a
+                        //         smarter solution atm.
+                        // todo@jldlaughlin: Improve Kinesis source activation #2195
+                        activator.activate_after(Duration::from_secs(5));
+                        return SourceStatus::Alive;
+                    }
+                }
+
+                for record in get_records_response.records {
+                    // For now, use the system's current timestamp to downgrade
+                    // capabilities.
+                    // todo: Implement better offset tracking for Kinesis sources #2219
+                    let cur = generate_timestamp();
+                    if cur > *cap.time() {
+                        cap.downgrade(&cur);
+                    } else {
+                        warn!(
+                            "{}: Unexpected - Kinesis source capability ahead of current system time ({} > {})",
+                            name, *cap.time(), cur
+                        );
+                    }
+
+                    let data = record.data.as_ref().to_vec();
+                    output.session(&cap).give((data, None))
+                }
+
+                if timer.elapsed().as_millis() > 100 {
+                    // We didn't drain the entire queue, so indicate that we
+                    // should run again. We suppress the activation when the
+                    // queue is drained, as in that case librdkafka is
+                    // configured to unpark our thread when a new message
+                    // arrives.
+                    activator.activate();
+                    return SourceStatus::Alive;
+                }
+                // Each Kinesis shard can support up to 5 read requests per second.
+                // This will throttle ourselves.
+                thread::sleep(Duration::from_millis(200));
```

Comment:
Eep! If you `thread::sleep` inside of a Timely dataflow worker, you will prevent that worker from doing other useful work and tank performance. The pattern instead is something like

```
if last_activated.since() < Duration::from_secs(5) {
    activator.active();
    return SourceStatus::Alive;
}
```

Can explain more in person if you'd like!